File: src/Commands/ProcessReaders.php
<?php

namespace Hoks\NewsRecommendation\Commands;

use Illuminate\Console\Command;
use Illuminate\Support\Facades\Redis;
use Hoks\NewsRecommendation\Models\UserMongo;
use Hoks\NewsRecommendation\Models\ArticleMongo;

class ProcessReaders extends Command
{
    /**
     * The name and signature of the console command.
     *
     * @var string
     */
    protected $signature = 'process:readers';

    /**
     * The console command description.
     *
     * @var string
     */
    protected $description = 'Proces redis readers';

    public function handle()
    {
        if(true){
            $output = shell_exec('/usr/bin/python3 '.__DIR__.'/process_readers.py');
        }else{
    
            //get todays date
            $todaysDate = now()->format('d-m-Y');

            $redisKeys = Redis::keys(config('newsrecommendation.redis_reader_prefix').'reader-'.'*');
            //set counter if we are going to limit evaluated users
            if(config('newsrecommendation.limit_users')){
                $counter = 0;
            }
            foreach ($redisKeys as $redisKey) {
                
                
                // //process only redis keys starting with "reader-"
                // if (strstr($redisKey, config('newsrecommendation.redis_reader_prefix').'reader-')) {
                    //check if we reached evaluated users limit
                    if(config('newsrecommendation.limit_users')){
                        if($counter >= config('newsrecommendation.limit_users_count')){
                            break;
                        }
                    }

                    //get articles ids string and convert it to array
                    $articlesString = Redis::get($redisKey);
                    $articlesIdsArray = explode('|', $articlesString);
                    $articlesIdsArray = array_unique($articlesIdsArray);

                    //we will make one big array with following structure
                    // 'tag_name' => nummber of occurences
                    $articleTagsSummedArray = [];
                    foreach ($articlesIdsArray as $articleId) {
                        //get every article tags from mongoDB
                        $articleMongo = ArticleMongo::where('article_id', (int) $articleId)->first();
                        if(isset($articleMongo) && !empty($articleMongo)) {
                            $articleTags = [];
                            $articleTags = $articleMongo->tags;
                            $siteId = $articleMongo->site_id;
                            $siteKey = 'site_'.$siteId;
                            if(isset($articleTagsSummedArray[$siteKey])){
                                $articleTagsSummedArray[$siteKey] = array_merge($articleTagsSummedArray[$siteKey], $articleTags);
                            }else{
                                $articleTagsSummedArray[$siteKey] = $articleTags;
                            }
                        }
                    }
                    
                    /**
                     * we now count duplicates to get something like this
                     * [
                     *      'tagA' => 10,
                     *      'tagB' => 7,
                     *      'tagC' => 5,
                     *      'tagD' => 3,
                     *      'tagE' => 1,
                     * ]
                     * and this array we actually put into db
                     */
                    $tagOccurencesArray = [];
                    foreach($articleTagsSummedArray as $key => $articleTagsSummedArrayOneSite){
                        $tagOccurencesArray[$key] = array_count_values($articleTagsSummedArrayOneSite);
                    }
                    //sorting array by the number of tags in descending order
                    $tagsOcurrencesSortArray = [];
                    foreach($tagOccurencesArray as $siteKey => $tagsOcurrences){
                        $tagsOcurrencesSort = [];
                        foreach($tagsOcurrences as $key => $value) {
                            $tagsOcurrencesSort[$key] = $value;
                        }
                        arsort($tagsOcurrencesSort);
                        $tagsOcurrencesSortArray[$siteKey] = $tagsOcurrencesSort;
                    }
                
                    //we create a user id
                    //there are two variants of redisKey, with and without firebase key
                    if(strstr($redisKey,'###')){
                        $longRedisKeyArray = explode('###',$redisKey);
                        $userId = str_replace(config('newsrecommendation.redis_reader_prefix').'reader-','',$longRedisKeyArray[0]);
                        $firebaseUid = $longRedisKeyArray[1];
                    }else{
                        $userId = str_replace(config('newsrecommendation.redis_reader_prefix').'reader-', '', $redisKey);
                        $firebaseUid = false;

                    }

                    if($firebaseUid){
                        //try to find if there is user with firebase_uid
                        $existingUser = UserMongo::where('firebase_uid', $firebaseUid)->orderBy('updated_at','desc')->first();
                        //if there is no user with firebase_uid and firebase_uid is in redis key, it could
                        //be the first login on anonymus user, so we try to find that user
                        if(!isset($existingUser) && empty($existingUser)){
                            $existingUser = UserMongo::where('user_id', $userId)->orderBy('updated_at','desc')->first();
                        }
                    }else{
                        $existingUser = UserMongo::where('user_id', $userId)->orderBy('updated_at','desc')->first();
                    }


                    if(isset($existingUser) && !empty($existingUser)) {
                        //we take all the tags that exist
                        $tags = $existingUser->tags;
                        //if there is a key with today
                        if(isset($tags[$todaysDate])) {
                            $userTagsArray = [];
                            $userTagsArray = (array) json_decode($existingUser->tags[$todaysDate]);
                            //todaysTags should contain all tags for todaysDate (both new and old)
                            $todaysTags = [];
                            foreach($userTagsArray as $siteKey => $userTagsOnSite){
                                //check if there are new tags for this site id
                                if(isset($tagsOcurrencesSortArray[$siteKey])){
                                    $userTagsOnSite = (array) $userTagsOnSite;
                                    //if there are already tags under today's key, add new ones
                                    if(count($userTagsOnSite) > 0) {
                                        //if new tag already exists, increase its occurence number in original array
                                        foreach ($userTagsOnSite as $key => $userTag) {
                                            if (isset($tagsOcurrencesSortArray[$siteKey][$key])) {
                                                $userTagsOnSite[$key] = $userTag + $tagsOcurrencesSortArray[$siteKey][$key];
                                                unset($tagsOcurrencesSortArray[$siteKey][$key]);
                                            }
                                        }
                                        //for this siteKey, merge new and old tags
                                        $userTagsArray[$siteKey] = array_merge($userTagsOnSite, $tagsOcurrencesSortArray[$siteKey]);
                                        $todaysTags[$siteKey] = $userTagsArray[$siteKey];
                                    }
                                    //if there is no key with today's date among the tags, create it and merge the new tags with the existing ones
                                    else {
                                        $todaysTags[$siteKey] = $tagsOcurrencesSortArray[$siteKey];
                                    }
                                }else{
                                    //if there are no new tags for some site, we just save old values for that site
                                    $todaysTags[$siteKey] = $userTagsArray[$siteKey];
                                }
                            }
                            $todaysTags =  [$todaysDate => json_encode($todaysTags)];
                            $existingUser->tags = array_merge($existingUser->tags,$todaysTags);
                        //if there is no key with today, create it and enter the tags
                        }else {
                            $todaysTags = [$todaysDate => json_encode($tagsOcurrencesSortArray)];
                            $existingUser->tags = array_merge($existingUser->tags,$todaysTags);
                        }

                        //if there are read news, add new ones, but keep the old ones
                        if(!empty($existingUser->read_news)) {
                            $readNewsOld = (array) json_decode($existingUser->read_news);
                            $readNewsNew = $articlesIdsArray;
                            $readNewsMerge = array_merge($readNewsOld, $readNewsNew);
                            $readNewsMerge = array_unique($readNewsMerge);
                            $readNewsMerge = array_values($readNewsMerge);
                            $existingUser->read_news = json_encode($readNewsMerge);
                        }else {
                            $existingUser->readed_news = json_encode($articlesIdsArray);
                        }

                        //we take all the tags from the user and format them as a multidimensional array
                        $userTagsAll = $existingUser->tags;
                        
                        $formatedTags = [];
                        foreach($userTagsAll as $date => $sitesTags) {
                            $sitesTags = (array) json_decode($sitesTags);
                            foreach($sitesTags as $siteKey => $tags){
                                $tags = (array) $tags;
                                $formatedTags[$date][$siteKey] = $tags;
                            }
                        }
                        //we check whether the user has read the news so that there is no duplication for the recommendation
                        $readNewsOld = (array) json_decode($existingUser->read_news);
                        if(empty($readNewsOld)) {
                            //we call a method that gives us an array of recommended news ids
                            if(config('newsrecommendation.use_weighted_algorithm')){
                                $recommendedArticles = $this->recommendedArticlesWeighted($formatedTags, $articlesIdsArray);
                            }else{
                                $recommendedArticles = $this->recommendedArticles($formatedTags, $articlesIdsArray);
                            }
                            $existingUser->news_recommendation = json_encode($recommendedArticles);
                        } else {
                            $readNewsMerge = array_merge($readNewsOld, $articlesIdsArray);
                            $readNewsMerge = array_unique($readNewsMerge);
                            if(config('newsrecommendation.use_weighted_algorithm')){
                                $recommendedArticles = $this->recommendedArticlesWeighted($formatedTags, $readNewsMerge);
                            }else{
                                $recommendedArticles = $this->recommendedArticles($formatedTags, $readNewsMerge);

                            }
                            $existingUser->news_recommendation = json_encode($recommendedArticles);
                        }
                        $existingUser->latest_update = $todaysDate;
                        //check if userId is changed (user used another device or deleted local storage)
                        if($firebaseUid){
                            if($userId != $existingUser->user_id){
                                $existingUser->user_id = $userId;
                            }
                            $existingUser->firebase_uid = $firebaseUid;
                        }
                        $existingUser->save();

                    //if there is no user, create one
                    }else {
                    //only create user if there are tags for articles (it may happen that user read some news that have no tags)
                    if(!empty($tagsOcurrencesSortArray)){
                            $userMongo = new UserMongo();
                            $userMongo->user_id = $userId;
                            $userMongo->read_news = json_encode($articlesIdsArray);
                            if(config('newsrecommendation.use_weighted_algorithm')){
                                $recommendedArticles = $this->recommendedArticlesWeighted([$todaysDate => $tagsOcurrencesSortArray], $articlesIdsArray);
                            }else{
                                $recommendedArticles = $this->recommendedArticles([$todaysDate => $tagsOcurrencesSortArray], $articlesIdsArray);
                            }
                            $userMongo->news_recommendation = json_encode($recommendedArticles);
                            $userMongo->tags = [$todaysDate => json_encode($tagsOcurrencesSortArray)];
                            $userMongo->latest_update = $todaysDate;
                            if($firebaseUid){
                                $userMongo->firebase_uid = $firebaseUid;
                            }
                            $userMongo->save();
                        }
                    }

                    //check if user has tags for day before n days, and delete it
                    if(isset($existingUser)) {
                        $daysKeepData = config('newsrecommendation.days_keep_data');
                        $dayToDeleteTag = now()->subDays($daysKeepData)->format('d-m-Y');

                        $tagsToDelete = $existingUser->tags;
                        if(isset($tagsToDelete[$dayToDeleteTag])) {
                            unset($tagsToDelete[$dayToDeleteTag]);
                            $existingUser->tags = $tagsToDelete;
                            $existingUser->save();
                        }

                    }

                    //delete redis key
                    Redis::command('DEL', [$redisKey]);
                    //increment counter if we are going to limit evaluated users
                    if(config('newsrecommendation.limit_users')){
                        $counter++;
                    }
                // }
                
            }
        }
    }


    /**
     * this method returns list of recommended articles ids for given user tags
     */
    protected function recommendedArticles($userTagsArray, $read_news = []){
        $exactTime = now();
        //we set values to integers
        if(!empty($read_news)) {
            $read_news = array_map('intval', $read_news);
        }
        //we merge tags for all days for one site in one element of an array
        $allTagsArray = [];
        //foreach trough days
        foreach($userTagsArray as $oneDayTagsArray){
            //foreach trough sites
            foreach($oneDayTagsArray as $siteKey => $oneDayOneSiteTags){
                //foreach trough tags for one day for one site
                $tempArray = [];
                foreach($oneDayOneSiteTags as $tag => $value){
                    if(isset($allTagsArray[$siteKey][$tag])){
                        $tempArray[$tag] = $allTagsArray[$siteKey][$tag]+$value;
                    }else{
                        $tempArray[$tag] = $value;
                    }
                }
                //sort tags by its occurrence
                arsort($tempArray);
                $allTagsArray[$siteKey] = $tempArray;
            }
            
        }
        //based on lowbar, we cut off part of array
        $arrayLength = config('newsrecommendation.tags_array_length');
        $tagsConstant = config('newsrecommendation.tags_constant');
        $recommendedArticlesCount = config('newsrecommendation.recommended_articles_count');
        $recommendedArticlesFinal = [];
        $excludedCategories = config('newsrecommendation.exclude_categories');
        $excludedSubcategories = config('newsrecommendation.exclude_subcategories');

        foreach($allTagsArray as $siteKey => $allTags){
            $allTags = array_slice($allTags,0,$arrayLength);
        
            //we sum all tags values (sum of all occurrences)
            $sumTagsValues = 0;
            foreach($allTags as $tag => $value){
                $sumTagsValues = $sumTagsValues+$value;
            }

            //we calculate tag coefficients
            $tagsCoefficients = [];
            foreach($allTags as $tag => $value){
                //here we get number that is less than 1 so we multiply by 10 and by constant from config
                $tagsCoefficients[$tag] = (int) round(($value/$sumTagsValues)*10*$tagsConstant);
            }
            //for each tag, we get tagCoeff number of articles (take just ids)
            $articles = [];
            foreach($tagsCoefficients as $tag => $value){
                //get $value articles with $tag, should be array of ids
                $siteIdForQuery = (int) str_replace('site_','',$siteKey);
                $articlesQuery = ArticleMongo::where('site_id',$siteIdForQuery)
                    ->where('tags', $tag)
                    ->whereNotIn('article_id', $read_news)
                    ->where('published',1)
                    ->where('publish_at','<=',$exactTime);
                
                if(isset($excludedCategories[$siteKey]) && !empty($excludedCategories[$siteKey])){
                    $articlesQuery = $articlesQuery->whereNotIn('category',$excludedCategories[$siteKey]);
                }
                if(isset($excludedSubcategories[$siteKey]) && !empty($excludedSubcategories[$siteKey])){
                    $articlesQuery = $articlesQuery->whereNotIn('subcategory',$excludedSubcategories[$siteKey]);
                }


                $articlesQuery = $articlesQuery->orderBy('publish_at', 'desc')
                    ->limit($value)
                    ->pluck('article_id');
                $articlesTemp = [];
                foreach($articlesQuery as $item){
                    $articlesTemp[] = $item;
                }
                $articles[$tag] = $articlesTemp;
            }
            //remove duplicates, keep originals (original is first of its kind)
            $seen = [];
            foreach($articles as $tag => $articlesArray){
                foreach($articlesArray as $key => $id){
                    if(in_array($id,$seen)){
                        unset($articles[$tag][$key]);
                    }
                    $seen[] = $id;
                }
                $articles[$tag] = array_values($articles[$tag]);

            }

            //we now have some articles for every tag
            //once again we recalculate how many of each we should take
            $articlesSum = 0;
            foreach($articles as $tag => $articlesArray){
            $articlesSum = $articlesSum + count($articlesArray);
            }
            $articlesCoefficient = [];
            if($articlesSum != 0) {
                foreach($articles as $tag => $articlesArray){
                    $articlesCoefficient[$tag] = (int) round((count($articlesArray)/$articlesSum)*$recommendedArticlesCount);
                }
            }
            // finnaly, we take recommended articles by the articlesCoeff
            $recommendedArticles = [];
            $enoughArticles = false;
            foreach($articlesCoefficient as $tag => $value){
                for ($x = 0; $x < $value; $x++) {
                    //break when you collect enough articles
                    if(count($recommendedArticles) >= $recommendedArticlesCount){
                        $enoughArticles = true;
                        break;
                    }
                    if(isset($articles[$tag][$x])){
                        $recommendedArticles[] = $articles[$tag][$x];
                        unset($articles[$tag][$x]);
                    }
                }
            }

            //if there are not enough articles, add some more
            if(!$enoughArticles){
                foreach($articles as $tag => $articlesArray){
                    foreach($articlesArray as $article){
                        if(count($recommendedArticles) >= $recommendedArticlesCount){
                        break;
                }
                        $recommendedArticles[] = $article;
                    }
                
                }
            }

            $recommendedArticlesFinal[$siteKey] = $recommendedArticles;

        }

        return $recommendedArticlesFinal;
        
    }

    /**
     * recommendations are determined using weighted matrix algorithm
     */
    public function recommendedArticlesWeighted($userTagsArray, $read_news = []){
        //we set values to integers
        if(!empty($read_news)) {
            $read_news = array_map('intval', $read_news);
        }
        //we merge tags for all days for one site in one element of an array
        $allTagsArray = [];
        //foreach trough days
        foreach($userTagsArray as $oneDayTagsArray){
            //foreach trough sites
            foreach($oneDayTagsArray as $siteKey => $oneDayOneSiteTags){
                //foreach trough tags for one day for one site
                $tempArray = [];
                foreach($oneDayOneSiteTags as $tag => $value){
                    if(isset($allTagsArray[$siteKey][$tag])){
                        $tempArray[$tag] = $allTagsArray[$siteKey][$tag]+$value;
                    }else{
                        $tempArray[$tag] = $value;
                    }
                }
                //sort tags by its occurrence
                arsort($tempArray);
                $allTagsArray[$siteKey] = $tempArray;
            }
            
        }
        //based on lowbar, we cut off part of array
        $arrayLength = config('newsrecommendation.tags_array_length');
        $recommendedArticlesCount = config('newsrecommendation.recommended_articles_count');
        $recommendedArticlesFinal = [];
        $excludedCategories = config('newsrecommendation.exclude_categories');
        $excludedSubcategories = config('newsrecommendation.exclude_subcategories');

        foreach($allTagsArray as $siteKey => $allTags){
            $allTags = array_slice($allTags,0,$arrayLength);
            //takes first tags_array_length  tags
            //get summed weighted matrix 
            $articlesQuery = ArticleMongo::where('published',1)->whereNotIn('article_id',$read_news);
            if(isset($excludedCategories[$siteKey]) && !empty($excludedCategories[$siteKey])){
                $articlesQuery = $articlesQuery->whereNotIn('category',$excludedCategories[$siteKey]);
            }
            if(isset($excludedSubcategories[$siteKey]) && !empty($excludedSubcategories[$siteKey])){
                $articlesQuery = $articlesQuery->whereNotIn('subcategory',$excludedSubcategories[$siteKey]);
            }
            $allOtherArticles = $articlesQuery->whereBetween('publish_at',[now()->subDays(config('newsrecommendation.weighted_algorithm_days')),now()])->get();


            $weightedArticlesValues = [];
            foreach($allOtherArticles as $article){
                $articleTags = $article->tags;
                foreach($allTags as $key => $count){
                    if(in_array($key,$articleTags)){
                        if(isset($weightedArticlesValues[$article->id])){
                            $weightedArticlesValues[$article->article_id] += $count;
                        }else{
                            $weightedArticlesValues[$article->article_id] = $count;
                        }
                    }
                }
            }

            arsort($weightedArticlesValues);

            $recommendedArticles = [];
            $counter = 0;

            foreach($weightedArticlesValues as $key => $articleItem){
                if($counter > $recommendedArticlesCount-1){
                    break;
                }
                $recommendedArticles[] = $key;
                $counter++;
            }

            $recommendedArticlesFinal[$siteKey] = $recommendedArticles;

        }

        return $recommendedArticlesFinal;
    }

   

}

File: src/Commands/process_readers.py
import json #hanldes JSON data
import redis #interacts with Redis
import pymongo # interacts with mongo
import threading
import os #used to get env and file paths
from concurrent.futures import ThreadPoolExecutor #manages multi-threaded execution
from datetime import datetime, timedelta #used to work with timestamps and dates
from collections import Counter, defaultdict #helps with counting occurences of elements

# Get the current script directory.
script_dir = os.path.dirname(os.path.abspath(__file__))
# Navigate from vendor/davor/news-recommendation/src/Commands to the Laravel base directory.
# Assuming your project's base is 4 levels up:
base_path = os.path.abspath(os.path.join(script_dir,"..", "..", "..", "..", ".."))
# Construct the path to python_recommendations_config.json in the Laravel config directory.
config_path = os.path.join(base_path, "config", "python_recommendations_config.json")


# Load configuration into config variable
with open(config_path, "r") as config_file:
    config = json.load(config_file)  # Ensure this loads a dictionary


# Redis configuration from .env
redis_host = os.getenv('REDIS_HOST', '127.0.0.1')  # Default to localhost if not set
redis_port = int(os.getenv('REDIS_PORT', 6379))    # Default to 6379 if not set
redis_password = os.getenv('REDIS_PASSWORD', None)  # Default to no password if not set
redis_db = int(os.getenv('REDIS_DB', 0))           # Default to DB 0 if not set
redis_prefix = os.getenv('REDIS_PREFIX', '')      # Default to DB 0 if not set


# MongoDB configuration from .env
mongo_host = os.getenv('DB_HOST_MONGO', 'mongo')  # Mongo host as per your env file
mongo_port = int(os.getenv('DB_PORT_MONGO', 27017))  # Mongo port from your .env file
mongo_db_name = os.getenv('DB_DATABASE_MONGO', 'mongo_test')  # Mongo database name
mongo_user = os.getenv('DB_USERNAME_MONGO', 'root')  # Mongo username
mongo_password = os.getenv('DB_PASSWORD_MONGO', 'cubes')  # Mongo password

# Redis connection
redis_client = redis.Redis(
   redis_host,
    redis_port,
    decode_responses=True
)
# MongoDB connection
mongo_client = pymongo.MongoClient(f"mongodb://{mongo_user}:{mongo_password}@" + mongo_host)
db = mongo_client[mongo_db_name]
user_collection = db[config["users_collection"]]
articles_collection = db[config["articles_collection"]]

# Get today's date
todays_date = datetime.now().strftime("%d-%m-%Y")


def process_reader(redis_key):
    # get value bahind redis key
    articles_string = redis_client.get(redis_key)
    # get article ids
    articles_ids_array = list(set(articles_string.split('|'))) if articles_string else [] #split explodes string, set removes duplicates, list converts it to list
    # Dictionary to store tags count per site
    article_tags_summed_dict = defaultdict(list)
    # Fetch article tags from MongoDB and organize by site
    for article_id in articles_ids_array:
        article_mongo = articles_collection.find_one({"article_id": int(article_id)})
        if article_mongo and 'tags' in article_mongo:
            site_key = f"site_{article_mongo['site_id']}"
            article_tags_summed_dict[site_key].extend(article_mongo['tags'])
            
    #get user and firebase uid
    user_id, firebase_uid = extract_user_id(redis_key)
    # check if we have user with those ids
    existing_user = find_user(user_id, firebase_uid)
    if existing_user:
        update_existing_user(existing_user, articles_ids_array)
    else:
        create_new_user(user_id, firebase_uid, articles_ids_array)

    cleanup_old_data(user_id)
    redis_client.delete(redis_key)


def extract_user_id(redis_key):
    # Extract user id and firebase uid if exists    
    # Check if the Redis key contains '###' it means we have firebase uid
    if "###" in redis_key:
        parts = redis_key.split("###")
        user_id = parts[0].replace(config["redis_reader_prefix"] + "reader-", "")
        firebase_uid = parts[1]  # Firebase UID exists
    else:
        user_id = redis_key.replace(config["redis_reader_prefix"] + "reader-", "")
        firebase_uid = None  # No Firebase UID in this case

    return user_id, firebase_uid



def find_user(user_id, firebase_uid):
    # find if we have user with any of the credentials
    query = {"firebase_uid": firebase_uid} if firebase_uid else {"user_id": user_id}
    return user_collection.find_one(query)

def update_existing_user(user, articles_ids):
    # Update tags occurrences
    tags_occurrences = count_tag_occurences(articles_ids)

    # Load existing user tags
    user_tags = json.loads(user.get("tags", "{}"))

    # Merge new tags into the existing structure
    if todays_date in user_tags:
        existing_tags = json.loads(user_tags[todays_date])
        for site, new_tags in tags_occurrences.items():
            existing_tags[site] = {**existing_tags.get(site, {}), **new_tags}
        user_tags[todays_date] = json.dumps(existing_tags, ensure_ascii=False)
    else:
        user_tags[todays_date] = json.dumps(tags_occurrences, ensure_ascii=False)

    # Serialize user tags properly
    user["tags"] = json.dumps(user_tags, ensure_ascii=False)

    # Update read news
    existing_read_news = set(map(int, json.loads(user.get("read_news", "[]"))))
    all_read_news = existing_read_news | set(map(int, articles_ids))
    user["read_news"] = json.dumps(list(all_read_news))  # Ensure JSON string format

    # Generate recommendations
    recommendations = recommended_articles_weighted(user_tags, list(all_read_news))
    user["news_recommendation"] = json.dumps(recommendations, ensure_ascii=False)

    # Update timestamps
    user["latest_update"] = todays_date
    user["updated_at"] = datetime.utcnow()

    # Save to MongoDB
    user_collection.update_one({"_id": user["_id"]}, {"$set": user})

def create_new_user(user_id, firebase_uid, articles_ids):
    # Count tag occurrences
    tags_occurrences = count_tag_occurences(articles_ids)

    if not tags_occurrences:
        return

    # Format tags correctly
    formatted_tags = json.dumps({todays_date: json.dumps(tags_occurrences, ensure_ascii=False)}, ensure_ascii=False)

    new_user = {
        "user_id": user_id,
        "firebase_uid": firebase_uid,
        "read_news": json.dumps(articles_ids),
        "tags": formatted_tags,
        "news_recommendation": json.dumps(recommended_articles_weighted({todays_date: tags_occurrences}, articles_ids), ensure_ascii=False),
        "latest_update": todays_date,
        "created_at": datetime.utcnow(),
        "updated_at": datetime.utcnow(),
    }
    
    user_collection.insert_one(new_user)



def count_tag_occurences(articles_ids):
    articles_ids = [int(id) for id in articles_ids]
    tags_count = {}

    # Fetch articles from MongoDB
    articles = articles_collection.find({"article_id": {"$in": articles_ids}}, {"tags": 1, "site_id": 1})

    for article in articles:
        site_key = f"site_{article['site_id']}"
        tags_count.setdefault(site_key, {})

        for tag in article.get("tags", []):
            tags_count[site_key][tag] = tags_count[site_key].get(tag, 0) + 1

    # Sort tags and return properly formatted dictionary
    sorted_tags_count = {
        site_key: dict(sorted(tags.items(), key=lambda x: x[1], reverse=True))
        for site_key, tags in tags_count.items()
    }

    return sorted_tags_count



def recommended_articles_weighted(user_tags_array, read_news=None):
    read_news = set(map(int, read_news)) if isinstance(read_news, list) else set(map(int, json.loads(read_news)))
    all_tags_dict = defaultdict(Counter)

    for one_day_tags in user_tags_array.values():
        if isinstance(one_day_tags, str):
            one_day_tags = json.loads(one_day_tags)
        for site_key, site_tags in one_day_tags.items():
            all_tags_dict[site_key].update(site_tags)

    array_length = config["tags_array_length"]
    recommended_articles_count = config["recommended_articles_count"]
    excluded_categories = config["exclude_categories"]
    excluded_subcategories = config["exclude_subcategories"]
    days_limit = config["weighted_algorithm_days"]
    date_treshold = datetime.utcnow() - timedelta(days=days_limit)
    recommended_articles_final = {}

    for site_key, tags in all_tags_dict.items():
        sorted_tags = dict(tags.most_common(array_length))

        query = {
            "published": 1,
            "article_id": {"$nin": list(read_news)},
            "publish_at": {"$gte": date_treshold}
        }

        if site_key in excluded_categories:
            query["category"] = {"$nin": excluded_categories[site_key]}
        if site_key in excluded_subcategories:
            query["subcategory"] = {"$nin": excluded_subcategories[site_key]}

        articles = articles_collection.find(query, {"article_id": 1, "tags": 1})
        weighted_articles = defaultdict(int)

        for article in articles:
            article_tags = article.get("tags", [])
            for tag in sorted_tags:
                if tag in article_tags:
                    weighted_articles[article["article_id"]] += sorted_tags[tag]

        recommended_articles = [
            article_id for article_id, _ in sorted(weighted_articles.items(), key=lambda item: item[1], reverse=True)[:recommended_articles_count]
        ]
        
        recommended_articles_final[site_key] = recommended_articles

    return recommended_articles_final


def cleanup_old_data(user_id):
    #delete old tags for user
    retention_days = config["days_keep_data"]
    old_date = (datetime.now() - timedelta(days=retention_days)).strftime("%d-%m-%Y")

    user = user_collection.find_one({"user_id": user_id})
    if user:
        user_tags = user.get("tags", {})
        if old_date in user_tags:
            del user_tags[old_date]
            user_collection.update_one({"_id": user["_id"]}, {"$set": {"tags": user_tags}})


def main():
    """Main function to fetch and process readers from Redis using multithreading."""
    redis_keys = redis_client.keys(config["redis_reader_prefix"] + "reader-*")
    max_users = config.get("limit_users", len(redis_keys))
    with ThreadPoolExecutor(max_workers=config["max_workers"]) as executor:
        executor.map(process_reader, redis_keys[:max_users])


if __name__ == "__main__":
    main()
